### PR TITLE
Fix SSL configs for forcing client connections

### PR DESCRIPTION
**Pull request by NyakudyaA**
- Default behaviour does not force SSL connections
- Connect to SSL using the default snake oil certificates
- SSL verification when using user-defined certificates

### CODE REVIEW

**Comment by NyakudyaA** [comment]
@lucernae Can you review this. As for the test can you follow the logic here https://github.com/kartoza/docker-postgis/pull/307/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R431

File: README.md [hunk 1]
```diff
@@ -400,9 +400,62 @@ See [the postgres documentation about SSL](https://www.postgresql.org/docs/11/li
...
+need to copy the `SSL_CA_FILE` file to the location `/home/$user/.postgresql/root.crt`
+or define an environment variable pointing to location of the `SSL_CA_FILE`
+example: `PGSSLROOTCERT=/etc/letsencrypt/root.crt`
+
+#### SSL connection inside the docker container using openssl certificates
+
+
+Generate the certificates inside the container
+```
+LETSENCRYPT_CERT_DIR=/etc/letsencrypt
```
Comment (timlinux) [inline README.md]: Does this make sense? We are not actually using letsencrypt in this case right?

**Comment by timlinux** [comment]
@NyakudyaA So I think we are still not in sync here. It would be good to reflect the following in the README too:

---------

There are three modalities in which you can work with SSL:

1. Optional: using the shipped snakeoil certificates
2. Forced SSL: forced using the shipped snakeoil certificates
3. Forced SSL with Certifcate Exchange: using SSL certificates signed by a certificate authority

In the first you can optionally set your client to use an encrypted connection, caveats outlined elsewhere in the README notwithstanding.

pg_hba.conf entries will be in the form:

```host    all             all             0.0.0.0/0               md5```

In the second scenario, you want to force all connections over SSL. If you have your own SSL certs deployed, they will be used, alternatively the snakeoil ones shipped with the container will be used. No CA cert will be required.

pg_hba.conf entries will be in the form:

```hostssl    all             all             0.0.0.0/0               md5 clientcert=0```

In the third scenario (most secure), you want to force all connections over SSL and you only want to accept SSL certificates that have been signed by a certificate authority.

```hostssl   all             all             0.0.0.0/0              cert clientcert=1```

---------

From what I can see in your PR we still miss support for option 2 - or did I miss something?

File: README.md (hunk 1 above)
Comment (NyakudyaA) [reply README.md]: This is just an example of how you could use the option you mentioned in the comment `Forced SSL with Certificate Exchange`

Comment (NyakudyaA) [reply README.md]: I will change the directory name to be generic

**Comment by NyakudyaA** [comment]
> @NyakudyaA So I think we are still not in sync here. It would be good to reflect the following in the README too:
> 
> There are three modalities in which you can work with SSL:
> 
>     1. Optional: using the shipped snakeoil certificates
> 
>     2. Forced SSL: forced using the shipped snakeoil certificates
> 
>     3. Forced SSL with Certifcate Exchange: using SSL certificates signed by a certificate authority
> 
> 
> In the first you can optionally set your client to use an encrypted connection, caveats outlined elsewhere in the README notwithstanding.
> 
> pg_hba.conf entries will be in the form:
> 
> `host all all 0.0.0.0/0 md5`

If you run the following

```
docker run --name "postgis"  -p 25432:5432 -d -t kartoza/postgis:13-3-1
```
 the pg_hba_entries will be `host all all 0.0.0.0/0 scram-sha-256`
> 
> In the second scenario, you want to force all connections over SSL. If you have your own SSL certs deployed, they will be used, alternatively the snakeoil ones shipped with the container will be used. No CA cert will be required.
> 
> pg_hba.conf entries will be in the form:
> 
> `hostssl all all 0.0.0.0/0 md5 clientcert=0`


if you run the following
 
```
docker run --name "postgis" -e FORCE_SSL=TRUE  -p 25432:5432 -d -t kartoza/postgis:13-3-1
```

The pg_hba.conf entry will be

`hostssl all all 0.0.0.0/0 scram-sha-256 clientcert=0`
> 
> In the third scenario (most secure), you want to force all connections over SSL and you only want to accept SSL certificates that have been signed by a certificate authority.
> 
> `hostssl all all 0.0.0.0/0 cert clientcert=1`

If you run the following
```
docker run -p 25433:5432 -e FORCE_SSL=TRUE -e SSL_CERT_FILE='/etc/certificates/fullchain.pem' -e SSL_KEY_FILE='/etc/certificates/privkey.pem' -e SSL_CA_FILE='/etc/certificates/root.crt' -v /tmp/postgres/setup.sh:/docker-entrypoint-initdb.d/setup.sh --name ssl -d kartoza/postgis:13-3.1 
```

the pg_hba.conf entry will be
`hostssl all all 0.0.0.0/0 cert`

When you set it to `cert` it implies directly that `clientcert=1` as cert only works with SSL Mode `verify-ca` and `verify-full`
> 
> From what I can see in your PR we still miss support for option 2 - or did I miss something?

All the options you outlined below are satisfied. scram-sha-256 is the default PASSWORD_AUTHENTICATION in all our images > 13

File: README.md (hunk 1 above)
Comment (timlinux) [reply README.md]: Thanks

**Comment by timlinux** [comment]
Thanks for the clarifications @NyakudyaA sounds perfect. Can you work those different scenarios into the readme? It will help make things much clearer.

**Pull request merged**